frontend-react/src/config/api/RSEndpoint.ts

```diff
@@ -0,0 +1,296 @@
+    apiHeaders = { ...apiHeaders, ...headers };
+
+    return apiHeaders;
+}
+
+const { API_ROOT } = config;
+
+// odd that there isn't already a useable implementation of this somewhere
+// or is there? where should this live?
+export const HTTPMethods = {
+    GET: "GET",
+    POST: "POST",
+    PUT: "PUT",
+    DELETE: "DELETE",
+    PATCH: "PATCH",
+} as const;
+export type HTTPMethod = (typeof HTTPMethods)[keyof typeof HTTPMethods];
+
+/**
+ * Copied from react-router for internal use with Endpoints.
+ * Please use react-router's types directly for react-router
+ * work.
+ */
+export type ParamParseSegment<Segment extends string> =
+    Segment extends `${infer LeftSegment}/${infer RightSegment}`
+        ? ParamParseSegment<LeftSegment> extends infer LeftResult
+            ? ParamParseSegment<RightSegment> extends infer RightResult
+                ? LeftResult extends string
+                    ? RightResult extends string
+                        ? LeftResult | RightResult
+                        : LeftResult
+                    : RightResult extends string
+                    ? RightResult
+                    : ParamParseFailed
+                : ParamParseFailed
+            : ParamParseSegment<RightSegment> extends infer RightResult
+            ? RightResult extends string
+                ? RightResult
+                : ParamParseFailed
+            : ParamParseFailed
+        : Segment extends `:${infer Remaining}`
+        ? Remaining
+        : ParamParseFailed;
+export type ParamParseKey<Segment extends string> =
+    ParamParseSegment<Segment> extends string
+        ? ParamParseSegment<Segment>
+        : string;
+/**
+ * The parameters that were parsed from the URL path.
+ */
+export type Params<Key extends string = string> = {
+    readonly [key in Key]: string;
+};
+export type ParamParseFailed = {
+    failed: true;
+};
+
+export interface EndpointConfig {
+    path: string;
+    methods: {
+        [key in HTTPMethod]?: unknown;
+    };
+    params?: unknown;
+    queryKey: string;
+}
+
+export interface EndpointConfigMap {
+    [k: string]: EndpointConfig;
+}
+
+export interface AxiosOptionsWithSegments
+    extends Omit<AxiosRequestConfig, "method" | "headers"> {
+    method: HTTPMethod;
+    segments?: StringIndexed<string>;
+    headers?: Partial<RawAxiosRequestHeaders>;
+}
+
+export type RSApiEndpoints = {
+    [endpointName: string]: RSEndpoint<any>;
+};
+
+export type RSEndpointMethodRequestType<
+    T extends RSEndpoint<any>,
+    M extends HTTPMethod
+> = T["meta"]["methods"][M] extends {}
+    ? NonNullable<T["meta"]["methods"][M]>
+    : never;
+
+export type RSEndpointFetchers<T extends RSEndpoint<any>> = {
+    [P in string & keyof T["meta"]["methods"] as Lowercase<P>]: (
+        args: Partial<Omit<RSOptionsWithSegments<T>, "method">>
+    ) => P extends HTTPMethod
+        ? Promise<RSEndpointMethodRequestType<T, P>>
+        : never;
+};
+
+export class RSEndpoint<E extends EndpointConfig = any> {
```

Comment:
Hmmm!  I've been mulling this over for a bit so forgive the wall of text that's inevitably incoming.

I'm wary about building more logic on top of RSEndpoint; this is starting to conflate the singular purpose of 1) what configuration is needed to make the request with 2) actually making the request -- and I'd argue that should be much more in the purview of react-query specifically.  I realize that the `queryFn` is likely intended to be used in `useQuery` calls as the fetcher, but I don't think we need to be adding fetcher logic to RSEndpoint.

My feeling is that there are a few concerns that can be handled separately:
- Generating request configuration, which includes fetch options (HTTP method, path interpolation) and react-query options (query key and query key prefixes, default values)
- Constructing the fetcher, which is just a function that calls Axios given a subset of the above
-  Wrapping `useQuery` and `useMutation` to accept custom parameters

For instance:

```js
// given a string url potentially with dynamic segments, return a string interpolated from a params object
// can also distinguish segment params versus query params
function getFullUrl(url, params) { ... }

// given request options, make a request via Axios
function createFetch({ 
    url, 
    method, // HTTP request method
    body, // optional body
    headers, // optional headers
    params  // optional query parameters
}) { ... }

// custom wrapper around useQuery given optional query config
function useResourceQuery({
    params,
    url,
    config, // optional resource-specific query options
    headers, // optional headers if needed on a per-endpoint level
    method, // can be omitted or defaulted to `GET` -- useQuery can be used for simple POSTs but unlikely in our usage (https://github.com/TanStack/query/discussions/1107)
}) { 
    // also may need to handle query parameters, depending on how getDynamicUrl works
    const fullUrl = getDynamicUrl(url, params);
    
    const query = useQuery({
        queryKey: [url, params],
        queryFn: () => createFetch({ url: fullUrl, method, headers }),
        config
    });

    return {
        ...query,
        // it'd also be nice to return some important things like queryKey and config.enable that aren't returned by default for whatever reason
        queryKey,
        isEnabled: config.enabled,
        ...
    };
}
```

We could also wrap useResourceQuery with another hook that'd handle authentication if needed:

```jsx
function useAuthenticatedResourceQuery({ ...otherStuff }) {
    const bearerToken = ...; // can get from LocalStorage or from SessionContext

    return useResourceQuery({ 
        ...otherStuff, 
        headers: { authorization: "Bearer blahblahblah" }
    });
}
```

And this composability would be useful to us in case if, say, we wanted to disable certain queries based on auth status:

```jsx
function useSenderOnlyResourceQuery({ ...otherStuff, config }) {
    const isOnlyPrimeAdminAndNotImpersonatingSender = ...; // can get from LocalStorage or from SessionContext

    return useResourceQuery({
        ...otherStuff,
        config: { ...config, enabled: !isOnlyPrimeAdminAndNotImpersonatingSender }
    });
}
```

I honestly don't think this is too far off from what's here; it's just a different, flatter approach with more emphasis on invoking functions as needed to accommodate flexible requirements rather than needing an upfront configuration.  However, I think it's important for us to keep our patterns more focused around the idea of composition to 1) play nicely with the very composition-forward React but also 2) to have better separation of concerns for more flexibility and easier debugging, testing, implementation, et cetera.

---

As an aside, I'm also wondering how useful RSEndpoint really is to us.  If we have a custom `useQuery` and `useMutation`, we should allow them to accept custom parameters that we can run the path interpolation against as another first-class function -- in which case, we could really downgrade RSEndpoint into simple functions that return Axios config options 🤷  However, this is a separate conversation from the above!

--- 

ANYWAY, my brain is completely shot today so I'm going to stop there.  I'm happy to chat more about this tomorrow if you'd like!